repo-agent/repowatch/controllers/repowatch_controller.go

```diff
@@ -80,9 +136,48 @@ func NewGithubClient(ctx context.Context, k8sClient client.Client, repoWatch *re
 		githubConfig["email"] = string(secret.Data["email"])
 	}
 
-	ts := oauth2.StaticTokenSource(
-		&oauth2.Token{AccessToken: string(pat)},
-	)
+	clientID := os.Getenv("GITHUB_CLIENT_ID")
+	clientSecret := os.Getenv("GITHUB_CLIENT_SECRET")
+	refreshToken := string(secret.Data["refresh_token"])
+	expiryStr := string(secret.Data["expiry"])
+
+	var ts oauth2.TokenSource
+
+	if clientID != "" && clientSecret != "" && refreshToken != "" {
+		// Use refresh flow
+		oauthConf := &oauth2.Config{
+			ClientID:     clientID,
+			ClientSecret: clientSecret,
+			Endpoint:     githuboauth.Endpoint,
+		}
+
+		token := &oauth2.Token{
+			AccessToken:  string(pat),
+			RefreshToken: refreshToken,
+		}
+
+		if expiryStr != "" {
+			if expiry, err := time.Parse(time.RFC3339, expiryStr); err == nil {
+				token.Expiry = expiry
+			}
+		}
+
+		// ReuseTokenSource will use the existing token if valid, or refresh it if expired.
+		// We wrap it in PersistingTokenSource to save the new token if refreshed.
+		reusingSource := oauthConf.TokenSource(ctx, token)
+		ts = &PersistingTokenSource{
```

Comment:
The `PersistingTokenSource` is initialized with the token from the secret, but there's no locking mechanism around the token refresh and persistence. If the token expires and multiple concurrent reconciles trigger a refresh, it could lead to a race condition where multiple goroutines attempt to refresh and update the secret simultaneously. Consider adding a mutex to `PersistingTokenSource` to ensure that token refresh and persistence is an atomic operation.

Reply:
Added a lock here